middle_end/flambda2/simplify_shared/closure_offsets.ml

```diff
@@ -689,145 +738,83 @@ module Greedy = struct
-            then offsets
-            else
-              Misc.fatal_errorf
-                "Env var %a is used in the current compilation unit, but not \
-                 present in the imported offsets."
-                Var_within_closure.print env_var
-          | Some info -> EO.add_env_var_offset offsets env_var info)
-      used_closure_vars offsets
-
-  let imported_and_used_offsets ~used_closure_ids ~used_closure_vars state =
-    match
-      (used_closure_ids, used_closure_vars : _ Or_unknown.t * _ Or_unknown.t)
-    with
-    | Known used_closure_ids, Known used_closure_vars ->
+  let imported_and_used_offsets ~used_names state =
+    match (used_names : _ Or_unknown.t) with
+    | Known
+        { closure_ids_normal;
+          closure_ids_in_types;
+          closure_vars_normal;
+          closure_vars_in_types
+        } ->
       state.used_offsets
-      |> collect_used_closure_ids state ~used_closure_ids
-      |> collect_used_closure_vars state ~used_closure_vars
-    | Unknown, Known _ | Known _, Unknown | Unknown, Unknown ->
-      EO.imported_offsets ()
-
-  (* Currently, it happens that some closure_ids/closure_vars can occur in
-     projections (and thus in the used_closure_ids/vars), but never occur in the
-     creation of a set of closures (and consequently have no slot, and thus are
-     not assigned any offset).
-
-     There are two main reasons why this could happen:
-
-     - we could have missed some sets of closures creations on the way up and
-     not recorded the associated constraints
-
-     - the closure_id/closure_var occur in dead code that was kept for some
-     reason (mainly due to the over-approximations made by data_flow, which
-     keeps alive more things than necessary). This regularly occurs when phantom
-     lets are generated, but can also occur without phantom let bindings.
-
-     Upon a missing offset for a closure_id/var, to_cmm will generate cmm
-     instructions that produce a segfault. This is safe in the second case, when
-     the closure_id/var with the missing offset occurs in dead code, but it is
-     incorrect in the first case.
-
-     The following check is intended to catch the first of these two cases, but
-     it cannot distinguish between the two cases, and this check results in a
-     lot of false positives. Thus we cannot always run this check. *)
-  let check_used_offsets state ~used_closure_ids ~used_closure_vars offsets =
-    if !Clflags.flambda_invariant_checks
-    then
-      match
-        (used_closure_ids, used_closure_vars : _ Or_unknown.t * _ Or_unknown.t)
-      with
-      | Known used_closure_ids, Known used_closure_vars ->
+      |> collect_used_closure_ids closure_ids_normal
+      |> collect_used_closure_ids closure_ids_in_types
+      |> collect_used_closure_vars closure_vars_normal
+      |> collect_used_closure_vars closure_vars_in_types
+    | Unknown -> EO.imported_offsets ()
```

Comment:
how would it be incorrect ? the reasoning for that line is: if we don't know precisely what is used in the current compilation unit, then let's re-export all the imported offsets, it might not be efficient, but, assuming we have offsets (in the imported offsets), for everything used, then we should be guaranteed that we export offsets for everything that occurs in the compilation unit.

Reply:
As far as I can tell it's missing the offsets for the current unit (for closure IDs and env vars that are local to this unit, and that should be present in `state.used_offsets`)

Reply:
You don't have to wait for the classic mode PRs, the free names are already computed correctly in `Closure_conversion` so you can use that.

Reply:
But I agree that it's correct to use `EO.imported_offsets ()` as a default; I was missing the fact that offset computations for the current unit occur after that.